### ISSUE TITLE
allow_lock_result_unwrap

### ISSUE THREAD

**Issue by glyn** [rust-lang/rust-clippy#16535]
### What it does

`allow_lock_result_unwrap` configures `unwrap_used` to allow a `std::sync::LockResult` to be unwrapped. The default value could be `false` for compatibility with the current behaviour or, potentially more usefully, `true`.

### Advantage

Avoids the need to `#[allow]` `unwrap()` after locking a `Mutex`.

### Drawbacks

None.

### Example

```rust
#[allow(clippy::unwrap_used)]
some_mutex
    .lock()
    .unwrap()
    ...;
```

Could be written as:

```rust
some_mutex
    .lock()
    .unwrap()
    ...;
```

### Comparison with existing lints

When the `unwrap_used` lint is applied, the user is forced to pepper their code with  `#[allow(clippy::unwrap_used)]` when Mutex lock results are unwrapped.

### Additional Context

The reasonableness of unwrapping `LockResult` is discussed here: https://www.reddit.com/r/rust/comments/10f6328/is_this_a_reasonable_way_to_handle_unwrapping/.

**Comment by ada4a** [comment]
Maybe a more future-proof version of this would be to provide a way to specify a list of error types that are okay to unwrap. See https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types for an example of a similar configuration option.

WDYT?

**Comment by glyn** [comment]
> Maybe a more future-proof version of this would be to provide a way to specify a list of error types that are okay to unwrap. See https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_types for an example of a similar configuration option.
> 
> WDYT?

Agreed. I've updated the OP. Thanks.

A simpler alternative might be `allow_unwrap_type` which would take a single type rather than a list and which could be used repeatedly if more than one types was needed. Would this alternative actually work though (or would each usage overwrite any previous ones)? If it did, this would avoid the minor edge case of `allow_unwrap_types` being used with an empty list.

**Comment by ada4a** [comment]
> Would this alternative actually work though (or would each usage overwrite any previous ones)?

Yeah I think it will get overwritten..

> If it did, this would avoid the minor edge case of `allow_unwrap_types` being used with an empty list.

I wouldn't say that's an edge case? One could even argue that that's we effectively have currently: no unwrap types are allowed, and so the lint fires on every `unwrap`.
